hbase-client/src/main/java/org/apache/hadoop/hbase/client/ConnectionImplementation.java

```diff
@@ -1123,12 +1127,26 @@ rpcControllerFactory, getMetaLookupPool(), connectionConfig.getMetaReadRpcTimeou
     }
   }
 
-  void takeUserRegionLock() throws IOException {
+  private long heldStartTime;
+
+  void takeUserRegionLock(int tries) throws IOException {
     try {
       long waitTime = connectionConfig.getMetaOperationTimeout();
+      long waitStartTime = 0;
+      if (metrics != null) {
+        metrics.updateUserRegionLockQueue(userRegionLock.getQueueLength());
+        waitStartTime = EnvironmentEdgeManager.currentTime();
+      }
       if (!userRegionLock.tryLock(waitTime, TimeUnit.MILLISECONDS)) {
+        if (metrics != null) {
+          metrics.incrUserRegionLockTimeout();
+        }
         throw new LockTimeoutException("Failed to get user region lock in" + waitTime + " ms. "
           + " for accessing meta region server.");
+      } else if (metrics != null) {
+        // successfully grabbed the lock, start timer of holding the lock
+        heldStartTime = EnvironmentEdgeManager.currentTime();
+        metrics.updateUserRegionLockWaiting(tries * waitTime + heldStartTime - waitStartTime);
```

Comment:
if we are here on a retry, it doesn't mean we had to wait `waitTime * tries` to get here. The first try could have happened without any time waiting for the lock. And if we waited `waitTime`, we would have thrown `LockTimeoutException` and we would not retry.
```suggestion
        metrics.updateUserRegionLockWaiting(heldStartTime - waitStartTime);
```

Reply:
Just directly measure and add up the time under lock.